### DISCUSSION TITLE
Do nRF bluetooth sniffers work with pyshark?

### DISCUSSION THREAD

**Discussion by damianjwilliams** [KimiNewt/pyshark#540] [Q&A]
I'm thinking of getting a nRF sniffer an ([Adafruit Bluefruit LE Friend](https://learn.adafruit.com/introducing-adafruit-ble-bluetooth-low-energy-friend/ble-sniffer))  and use it to read streaming advertising packets from a peripheral device.  It seems like pyshark/tshark did not work with with the board (at least a few years ago) e.g. [Stack Overflow](https://stackoverflow.com/questions/50977566/tshark-will-not-capture-btle-packets-with-nrf-sniffer) and [Nordic Semiconductor Q &A](https://devzone.nordicsemi.com/f/nordic-q-a/68233/unable-to-capture-ble-packets-using-nrf51-with-python-interface). Is that still the case?

**Comment by KimiNewt** [comment]
I'm not familiar with it, but the easiest way to find out is just to try running `tshark` on it. If it runs, pyshark should too.

**Reply by damianjwilliams** [reply]
Thanks for the reply. It seems to work with tshark but not with pyshark:

 Runnining tshark using Windows 10 command line
```
C:\Program Files\Wireshark>tshark -D
1. \Device\NPF_{96009682-775E-485A-825D-56EA8AAF568F} (Local Area Connection* 11)
2. \Device\NPF_{6EF0DAE2-F5D3-46D2-8F93-A827E04DA504} (Local Area Connection* 10)
3. \Device\NPF_{E2219539-786A-4D11-B728-BEDBC415D26B} (Local Area Connection* 9)
4. \Device\NPF_{C68B0061-6956-4D7A-8E69-5CFD402BAE9F} (Bluetooth Network Connection 2)
5. \Device\NPF_{DF97E23C-D9D3-4196-92C4-159FE9F14F6B} (Wi-Fi)
6. \Device\NPF_{9ED7B0A0-17BF-4DAF-B27E-501EF1492897} (Local Area Connection* 3)
7. \Device\NPF_{5563DB7F-6508-4E85-A846-1CBFC93B1F19} (Local Area Connection* 2)
8. \Device\NPF_Loopback (Adapter for loopback traffic capture)
9. COM14-3.6 (nRF Sniffer for Bluetooth LE COM14)
10. \\.\USBPcap1 (USBPcap1)

C:\Program Files\Wireshark>tshark -i "nRF Sniffer for Bluetooth LE COM14"
Capturing on 'nRF Sniffer for Bluetooth LE COM14'
 ** (tshark:1728) 13:10:12.102239 [Main MESSAGE] -- Capture started.
 ** (tshark:1728) 13:10:12.105879 [Main MESSAGE] -- File: "C:\Users\Damian\AppData\Local\Temp\wireshark_nRF Sniffer for Bluetooth LE COM14QLQDN1.pcapng"
    1   0.000000 Apple_ab:bd:6e → Broadcast    LE LL ADV_IND
    2   0.156217 5c:23:42:db:fe:f8 → Broadcast    LE LL ADV_SCAN_IND
    3   0.218738 5c:23:42:db:fe:f8 → Broadcast    LE LL ADV_SCAN_IND
    4   0.312428 59:09:60:44:e5:59 → Broadcast    LE LL ADV_IND
4 packets captured

```
but when I try to run pyshark:
```
>>> capture = pyshark.LiveCapture(interface="nRF Sniffer for Bluetooth LE COM14")
>>> capture.sniff(timeout=50)
>>> capture
<LiveCapture (0 packets)>

```
Interestingly, I can run tshark as a subprocess in Python:
```
import subprocess

tsharkCall = ["C:\\Program Files\\Wireshark\\tshark.exe", "-i","nRF Sniffer for Bluetooth LE COM14","-l"]
tsharkProc = subprocess.Popen(tsharkCall,
                              bufsize=0,
                              stdout=subprocess.PIPE)
for packet in iter(tsharkProc.stdout.readline, b" "):
    packet_string   = packet.rstrip().decode("utf-8") #bytes to string
    #packet_info
    print(packet_string)
```
Terminal output:
```
Capturing on 'nRF Sniffer for Bluetooth LE COM14'
 ** (tshark:12276) 13:28:25.254788 [Main MESSAGE] -- Capture started.
 ** (tshark:12276) 13:28:25.258882 [Main MESSAGE] -- File: "C:\Users\Damian\AppData\Local\Temp\wireshark_nRF Sniffer for Bluetooth LE COM14U2Z9M1.pcapng"
1     1   0.000000 4f:07:75:87:e1:e3 → Broadcast    LE LL ADV_IND
    2   0.156257 23:63:47:e0:e2:b3 → Broadcast    LE LL ADV_NONCONN_IND
    3   0.265636 4f:07:75:87:e1:e3 → Broadcast    LE LL ADV_IND
    4   0.359359 VSIndust_55:f3:3b → Broadcast    LE LL ADV_NONCONN_IND
```


Thanks,
Damian

**Reply by KimiNewt** [reply]
Can you first run the same thing only with `debug=True` in the `LiveCapture` constructor?

**Reply by damianjwilliams** [reply]
Hi,

Here you go:

```
capture = pyshark.LiveCapture(interface="COM14-3.6",debug=True)
capture.sniff(timeout=20)
2022-06-09 10:00:11,334 - LiveCapture - DEBUG - Creating Dumpcap subprocess with parameters: C:/Program Files/Wireshark/dumpcap.exe -q -i COM14-3.6 -w -
2022-06-09 10:00:11,343 - LiveCapture - DEBUG - Dumpcap subprocess (pid 16228) created
2022-06-09 10:00:11,344 - LiveCapture - DEBUG - Creating TShark subprocess with parameters: C:/Program Files/Wireshark/tshark.exe -l -n -T pdml -i -
2022-06-09 10:00:11,344 - LiveCapture - DEBUG - Executable: C:/Program Files/Wireshark/tshark.exe
2022-06-09 10:00:11,350 - LiveCapture - DEBUG - TShark subprocess (pid 5936) created
2022-06-09 10:00:11,350 - LiveCapture - DEBUG - Starting to go through packets
Capturing on 'COM14-3.6'
dumpcap: The capture session could not be initiated on capture device "COM14-3.6" (No such device exists).
Please check that you have the proper interface or pipe specified.
2022-06-09 10:00:18,685 - LiveCapture - DEBUG - Cleanup Subprocess (pid 16228)
2022-06-09 10:00:18,685 - LiveCapture - DEBUG - Cleanup Subprocess (pid 5936)
```

Looks like dumpcap is not recognizing the sniffer but tshark is. 

To confirm this, when I run dumpcap from the command line I get a list of interfaces missing the the sniffer 
```
C:\Program Files\Wireshark>dumpcap -D

5. \Device\NPF_{DF97E23C-D9D3-4196-92C4-159FE9F14F6B} (Wi-Fi)
6. \Device\NPF_{9ED7B0A0-17BF-4DAF-B27E-501EF1492897} (Local Area Connection* 3)
7. \Device\NPF_{5563DB7F-6508-4E85-A846-1CBFC93B1F19} (Local Area Connection* 2)
8. \Device\NPF_Loopback (Adapter for loopback traffic capture)
```
but when I run tshark, the sniffer is present
```
C:\Program Files\Wireshark>tshark -D

5. \Device\NPF_{DF97E23C-D9D3-4196-92C4-159FE9F14F6B} (Wi-Fi)
6. \Device\NPF_{9ED7B0A0-17BF-4DAF-B27E-501EF1492897} (Local Area Connection* 3)
7. \Device\NPF_{5563DB7F-6508-4E85-A846-1CBFC93B1F19} (Local Area Connection* 2)
8. \Device\NPF_Loopback (Adapter for loopback traffic capture)
9. COM14-3.6 (nRF Sniffer for Bluetooth LE COM14)
10. \\.\USBPcap1 (USBPcap1)

```

Does dumpcap need to be run in addition to tshark?

Thanks,

Damian

**Reply by KimiNewt** [reply]
Yes, for live captures dumpcap needs to run as well. It's really bizarre that not all the same interfaces are available. I assume you can see all of them in wireshark?

**Reply by damianjwilliams** [reply]
Yup, I can see all of the interfaces on Wireshark

**Reply by alereds94** [reply]
Hi all, 
I am trying to automate wireshark sniffing with Python. I experience the same issues reported in this topic. The only way I've found up to now is to launch tshark from a subprocess in Python, save the output to file with -w and then use a pyshark.FileCapture for later analysis.
Is there any way to implement a pyshark.LiveCapture directly with the "nRF Sniffer for Bluetooth LE" interface? Thanks in advance

**Reply by DaveDixon** [reply]
Ther are things you can see in wireshark/tshark you can't see in pyshark, because pyshark relies heavily on dumpcap, and, to quote the documentation:

"Unlike Wireshark and tshark, dumpcap cannot see non-physical interfaces like extcap interfaces. tshark has most of the same flags that dumpcap has because tshark calls dumpcap for much of its capture functionality." (from https://tshark.dev/capture/dumpcap/)

Thus, I can see my nRF Sniffer in tshark (tshark -D), and I can list it in python using pyshark.tshark.tshark.get_tshark_interfaces(), but capture.debug() says "dumpcap: There is no device named ..."

So, but putting the nRF Sniffer code in the extcap folder for wireshark, you're telling wireshark and tshark about it, but dumpcap can't see it because it's not physical.
